### DISCUSSION TITLE
Tell to Meson to compile C files with CXX compiler

### DISCUSSION THREAD

**Discussion by cprieto** [mesonbuild/meson#11603] [Q&A] [answered]
In certain cases, like for example, compiling Catch2 tests with a C language file, you need to tell the compiler to use the C++ compiler to compile the C files (instead of using the CC compiler).

In CMake is easily done with `set_source_files_properties` and setting the `LANGUAGE` to `CXX`. I had been searching in the documentation without any success, maybe I am missing something?

I tried already setting the language to `cpp` (it fails, no C compiler is found) and to both `cpp` and `c` and nope, no success.

**Comment by eli-schwartz** [comment]
Does the C language file contain C++ code defining the tests? If so, why not name it .cpp instead?

If it truly only contains C code but is needed in the binary along with a .cpp file containing the test definition, then is it not sufficient to link both the C functions and the C++ test definitions together using a C++ linker? Meson will do that automatically for targets that contain a mixture of C and C++ source files.

**Reply by cprieto** [reply]
Thanks, renaming the file is a good solution but it sounds more like a hack than actually solution (when compared with the file properties in CMake), is there a place where I can suggest this as a feature?

**Reply by eli-schwartz** [reply]
It is a hack to name C++ files with C++ file extensions rather than C file extensions?

An actual solution is to create new features that need to be manually maintained in the build system recipe (meson.build file) using custom syntax in order to pass hardcoded flags to the compiler telling it to override format detection?

I'm genuinely surprised to hear this. I've heard, once or twice, people argue that for legacy compat reasons they "can't" rename the files, but this is the first time I think I've ever heard someone say they find it more elegant to use the wrong filename.

**Reply by cprieto** [reply]
Ok, let's say it is not a hack, at the end of the day I can do it in CMake and not in Meson, so there is that... :man_shrugging: 

Now, as part of the conversation, if instead of confronting and discuss if that is a hack or not (which absolutely helps nobody) you have a reference about how to do it _without_ renaming it (for whatever reason it is) I will highly appreciate it.

**Reply by eli-schwartz** [reply]
Meson has no comparable feature for per-file properties.

**Reply by cprieto** [reply]
Thanks man, I just wanted to confirm it :smiley_cat:

**Comment by eli-schwartz** [comment]
Another option would be (useful if the files belong to some other project and you can't rename them) to `#include` the C file in a C++ file.

**Comment by eli-schwartz** [comment]
Another option would be to include a C language for the project, but pass `c_args: ['-x', 'c++']` to the target -- it will apply to all C files in that target. Assuming there are mixed C/C++ in that target it will already default to promoting the linker to be the C++ one, but you can set the link language manually too.

**Comment by black-sliver** [comment, marked as answer]
Is this still unsolved? (That is instructing Meson to use CXX for *.c)?
My use-case is very different from the original question: what I want to do is compile liblua as C++, which allows using C++ features in it, such as using Exceptions instead of longjmp.

`c_args: ['-x', 'c++']` works as a work-around, but this really is a hack because this requires you to define the C compiler as something that understands C++. It's also somewhat confusing that you have to define both `'c'` and `'c++'` for the project and `add_project_arguments()` to both languages.

**Comment by bonzini** [comment]
Meson generally prefers to use conventions over configuration, so for example the extension drives which compiler is used. Can you give an example of this use of liblua with exceptions? Is the use supported in Lua's sources?

**Reply by black-sliver** [reply]
See the comment in lua/ldo.c here https://github.com/lua/lua/blob/7579fc9d7ed90240487251dfb69168f8e64e9294/ldo.c#L69

**Reply by black-sliver** [reply]
The reason to use C++ Exceptions for Lua errors is that longjmp can mess up RAII (and C++ Exceptions can mess up the Lua state).

**Reply by bonzini** [reply]
If it is just one file, writing a single one-line file with a #include directive would be the easiest, to be honest. Even with simple makefiles you would have to use `-x c++`.

**Reply by black-sliver** [reply]
Yeah, but I would ask a C++ compiler with -x to treat input as C++, not a C compiler.

As long as I only support clang and gcc (and maybe msvc at some point), it doesn't really matter, but I am in the process of going from Makefile to Meson to reduce the amount of jank 😅

**Comment by Volker-Weissmann** [comment]
Maybe it helps you to know that meson will treat .c files as C code, but .C files as C++ code. 
But I should add that while I was the one who demanded and implemented it (#8747) I should say that I no longer think that is/was a good idea. I think renaming the files is the best solution, unless meson would add some kind of file-properties.

**Reply by black-sliver** [reply]
I don't see how that helps. If I could rename the file, I could as well rename it to .cpp.
